### PR TITLE
refactor: async with multiaddr conn

### PR DESCRIPTION
**Pull request by vasco-santos**
BREAKING CHANGE: Switch to using async/await and async iterators. The transport and connection interfaces have changed.

Compliant with the newest [ibp2p/interface-transport](https://github.com/libp2p/interface-transport)

Branched of #82 

I added a local test for using `wss` multiaddr, which was not tested in the older version. However, I created #93 to test `wss` using the `interface-transport` tests in the future.

Needs:

- [ ]  https://github.com/vasco-santos/it-ws/commit/0c7e43b68f1d40c2296804a0ee7ed093da2d261c

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,68 +1,135 @@
...
+    log('outbound connection %s upgraded', maConn.remoteAddr)
+    return conn
+  }
+
+  /**
+   * @private
+   * @param {Multiaddr} ma
+   * @param {object} [options]
+   * @param {AbortSignal} [options.signal] Used to abort dial requests
+   * @returns {Promise<Socket>} Resolves a TCP Socket
```
Comment (jacobheun) [inline src/index.js]: This is actually an extended duplex iterable isn't it? I wouldn't call it a TCP socket, Websocket would be fine although not entirely accurate.

File: src/socket-to-conn.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    async sink (source) {
+      if (options.signal) {
+        source = abortable(source, options.signal)
+      }
+
+      try {
+        await socket.sink(source)
+      } catch (err) {
+        if (err.type !== 'aborted') {
+          log(err)
```
Comment (jacobheun) [inline src/socket-to-conn.js]: Should we make this `log.error(err)`? Makes it easier to just target errors with debugging on.

File: src/socket-to-conn.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    localAddr: undefined,
+
+    // If the remote address was passed, use it - it may have the peer ID encapsulated
+    remoteAddr: options.remoteAddr || toMultiaddr(socket.remoteAddress, socket.remotePort),
+
+    timeline: { open: Date.now() },
+
+    close () {
+      return new Promise(async (resolve) => { // eslint-disable-line no-async-promise-executor
```
Comment (jacobheun) [inline src/socket-to-conn.js]: I think we should change the usage here so we're not using async inside the promise constructor.

We could use [p-timeout](https://github.com/sindresorhus/p-timeout), pass it `socket.close()` and await that.

File: test/node.js [hunk 4]
```diff
@@ -440,124 +417,35 @@ describe('filter addrs', () => {
   })
 })
 
-describe('valid Connection', () => {
-  const ma = multiaddr('/ip4/127.0.0.1/tcp/9092/ws')
+describe.skip('valid localAddr and remoteAddr', () => {
```
Comment (jacobheun) [inline test/node.js]: Why is this being skipped?

File: src/socket-to-conn.js (hunk 3 above)
Comment (vasco-santos) [reply src/socket-to-conn.js]: Cool, I found a bug thanks to this on my `it-ws` branch. Already updating it and changing this in favor of `p-timeout`

File: package.json [hunk 5]
```diff
@@ -40,21 +38,26 @@
   },
   "homepage": "https://github.com/libp2p/js-libp2p-websockets#readme",
   "dependencies": {
+    "abortable-iterator": "^2.1.0",
     "class-is": "^1.1.0",
     "debug": "^4.1.1",
-    "interface-connection": "~0.3.3",
-    "mafmt": "^6.0.7",
+    "err-code": "^2.0.0",
+    "it-ws": "vasco-santos/it-ws#feat/add-properties-and-functions-to-client-and-server",
```
Comment (vasco-santos) [inline package.json]: @jacobheun what is your favorite procedure here? Create a tag in my repo's branch?

Comment (jacobheun) [reply package.json]: Yeah, tags for dependencies that should eventually get updated. It avoids the volatility of a branch. It's not ideal, but publishing another npm module is overkill imo.

Comment (vasco-santos) [reply package.json]: ok, updated!

**Pull request merged**